File: 2.linked-list/single-linked-list/add-two-numbers/add_two_numbers.rb
class ListNode
  def initialize(val, next = nill)
    @val = val
    @next = next
  end
end

def add_two_numbers(l1, l2)
  dummy = cur = new ListNode(0)
  carry = 0
  while (l1 or l2 or carry)
    if l1
      carry += l1.val
      l1 = l1.next
    end
    if l2
      carry += l2.val
      l2 = l2.next
    end
    cur.next = new ListNode(carry%10)
    cur = cur.next
    carry /= 10
  end
  return dummy.next
end

File: 2.linked-list/single-linked-list/add-two-numbers/addTwoNumbers.js
class ListNode {
  constructor(val, next = null) {
    this.val = val;
    this.next = next;
  }
}
const addTwoNumbers = function (l1, l2) {
  let cur = new ListNode(0);
  let dummy = cur;
  let carry = 0;
  while (l1 || l2 || carry) {
    if (l1) {
      carry = l1.val;
      l1 = l1.next;
    }
    if (l2) {
      carry = l2.val;
      l2 = l2.next;
    }
    cur.next = new ListNode(carry % 10);
    cur = cur.next;
    carry /= 10;
  }
  return dummy.next;
};
